mlflow/store/sqlalchemy_store.py

```diff
@@ -0,0 +1,222 @@
+        self._save_to_db(experiment)
+
+    def create_run(self, experiment_id, user_id, run_name, source_type, source_name,
+                   entry_point_name, start_time, source_version, tags, parent_run_id):
+        _ = parent_run_id
+        experiment = self.get_experiment(experiment_id)
+
+        if experiment.lifecycle_stage != entities.Experiment.ACTIVE_LIFECYCLE:
+            raise MlflowException('Experiment id={} must be active'.format(experiment_id),
+                                  error_codes.INVALID_STATE)
+
+        run = models.SqlRun(name=run_name, artifact_uri=None,
+                            experiment_id=experiment_id, source_type=source_type,
+                            source_name=source_name, entry_point_name=entry_point_name,
+                            user_id=user_id, status=entities.RunStatus.RUNNING,
+                            start_time=start_time, end_time=None,
+                            source_version=source_version,
+                            lifecycle_stage=entities.RunInfo.ACTIVE_LIFECYCLE)
+
+        for tag in tags:
+            run.tags.append(models.SqlRunTag(key=tag.key, value=tag.value))
+        self._save_to_db([run])
+
+        run = run.to_mlflow_entity()
+
+        return run
+
+    def update_run_info(self, run_uuid, run_status, end_time):
+        run = self.session.query(models.SqlRun).filter_by(run_uuid=run_uuid).first()
+        run.status = run_status
+        run.end_time = end_time
+
+        self._save_to_db(run)
+        run = run.to_mlflow_entity()
+
+        return run.info
+
+    def restore_run(self, run_id):
+        run = self.session.query(models.SqlRun).filter_by(run_uuid=run_id).first()
+        run.is_deleted = False
+        self._save_to_db(run)
+
+    def get_run(self, run_uuid):
+        run = self.session.query(models.SqlRun).filter_by(run_uuid=run_uuid,
+                                                          is_deleted=False).first()
+        if run is None:
+            raise MlflowException('Run(uuid={}) doesn\'t exist'.format(run_uuid),
+                                  error_codes.RESOURCE_DOES_NOT_EXIST)
+
+        return run.to_mlflow_entity()
+
+    def delete_run(self, run_id):
+        run = self.session.query(models.SqlRun).filter_by(run_uuid=run_id).first()
+        run.is_deleted = True
+        self._save_to_db(run)
+
+    def log_metric(self, run_uuid, metric):
+        try:
+            self._get_or_create(models.SqlMetric, run_uuid=run_uuid, key=metric.key,
+                                value=metric.value, timestamp=metric.timestamp)
+        except IntegrityError:
+            raise MlflowException('Metric={} must be unique'.format(metric),
+                                  error_codes.INVALID_PARAMETER_VALUE)
+
+    def log_param(self, run_uuid, param):
+        # if we try to update the value of an existing param this will fail
+        # because it will try to create it with same run_uuid, param key
```

Comment:
Will this work in all types of sql database backends? Would be an error if it creates a new one that overwrites older one.